accumulator/invocation.go

```diff
@@ -0,0 +1,106 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package accumulator
+
+import (
+	"bytes"
+	"time"
+
+	"github.com/tidwall/gjson"
+	"go.elastic.co/fastjson"
+)
+
+// Invocation holds data for each function invocation and finalizes
+// the data when `platform.report` type log is received for the
+// specific invocation identified by request ID.
+type Invocation struct {
+	// RequestID is the id to identify the invocation.
+	RequestID string
+	// Timestamp is the time of the invocation.
+	Timestamp time.Time
+	// DeadlineMs is the function execution deadline.
+	DeadlineMs int64
+	// FunctionARN requested. Can be different in each invoke that
+	// executes the same version.
+	FunctionARN string
+	// TransactionID is the ID generated for a transaction for the
+	// current invocation. It is populated by the request from agent.
+	TransactionID string
+	// TraceID is the ID generated for a trace for the current invocation.
+	// It is populated by the request from agent.
+	TraceID string
+	// Metadata stripped data from the agent. Each line is represented as
+	// a seperate entry.
+	agentData [][]byte
+}
+
+// Finalize searches the agent data for an invocation to find the root txn
+// for the invocation. If root txn is not found then a new txn is created
+// with the payload submitted by the agent.
+func (inc *Invocation) Finalize(status string) {
+	if inc.TransactionID == "" {
+		return
+	}
+	if rootTxnIdx := inc.findRootTxn(); rootTxnIdx == -1 {
+		inc.createProxyTxn(status)
+	}
+}
+
+func (inc *Invocation) createProxyTxn(status string) int {
+	var w fastjson.Writer
+	w.RawString(`{"transaction":{"id":`)
+	w.String(inc.TransactionID)
+	w.RawString(`,"trace_id":`)
+	w.String(inc.TraceID)
+	w.RawString(`,"result":`)
```

Comment:
Should we also set `outcome`?

Reply:
What should be the value for the `outcome`? I was a bit confused between what these fields indicates.

Initially, I chose to set only `outcome`, as the description for `transaction.result` felt not completely aligned with the [description of the result field](https://www.elastic.co/guide/en/apm/guide/8.4/exported-fields-apm-transaction.html), however, I can only find `event.outcome` in the documentations and no `transaction.outcome` but we do seem to [export transaction.outcome from apm-agent-go](https://github.com/elastic/apm-agent-go/blob/4d07ba809e943acb94c347517cfbc1d760a65297/model/model.go#L276-L277).

Reply:
It would be set to the string "failure" for failed transactions. Otherwise I think we would omit it.

> Initially, I chose to set only outcome, as the [description for transaction.result](https://www.elastic.co/guide/en/apm/guide/8.4/exported-fields-apm-transaction.html) felt not completely aligned with what we were setting, however, I can only find event.outcome in the documentations and no transaction.outcome but we do seem to [export transaction.outcome from apm-agent-go](https://github.com/elastic/apm-agent-go/blob/4d07ba809e943acb94c347517cfbc1d760a65297/model/model.go#L276-L277).

`transaction.outcome` in intake-v2 results in the `event.outcome` field being set in docs. This field is used for calculating the rather course-grained "failed transaction rate" metric.

`transaction.result` is free-form, and can be used to group transactions by a more specific error reason. This is where we would set "timeout", "oom", etc.